src/contract.rs

```diff
@@ -0,0 +1,203 @@
+        msg.gov_module.to_string(),
+        &Roles::all_roles().into_iter().collect(),
+    )?;
+
+    execute::add_new_paths(&mut deps, msg.paths, env.block.time)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("ibc_module", msg.ibc_module.to_string())
+        .add_attribute("gov_module", msg.gov_module.to_string()))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    // check to see if special permissions are required to invoke the message, and that the sender has the required permissions
+    can_invoke_message(&deps, &info, &msg)?;
+    // check to see if messages sent by MessageInfo::sender require a timelock
+    //
+    // if a timelock is required the message must be queued for execution
+
+    if msg.skip_queue() {
+        match_execute(&mut deps, &env, msg)
+    } else if must_queue_message(&mut deps, &info) {
+        let message_id = queue_message(&mut deps, env, msg, info)?;
+        Ok(Response::new().add_attribute("message.id", message_id))
+    } else {
+        match_execute(&mut deps, &env, msg)
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn sudo(deps: DepsMut, env: Env, msg: SudoMsg) -> Result<Response, ContractError> {
+    match msg {
+        SudoMsg::SendPacket {
+            packet,
+            #[cfg(test)]
+            channel_value_mock,
+        } => sudo::process_packet(
+            deps,
+            packet,
+            FlowType::Out,
+            env.block.time,
+            #[cfg(test)]
+            channel_value_mock,
+        ),
+        SudoMsg::RecvPacket {
+            packet,
+            #[cfg(test)]
+            channel_value_mock,
+        } => sudo::process_packet(
+            deps,
+            packet,
+            FlowType::In,
+            env.block.time,
+            #[cfg(test)]
+            channel_value_mock,
+        ),
+        SudoMsg::UndoSend { packet } => sudo::undo_send(deps, packet),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetQuotas { channel_id, denom } => {
+            query::get_quotas(deps.storage, channel_id, denom)
+        }
+        QueryMsg::GetRoleOwners => query::get_role_owners(deps.storage),
+        QueryMsg::GetRoles { owner } => query::get_roles(deps.storage, owner),
+        QueryMsg::GetMessageIds => query::get_message_ids(deps.storage),
+        QueryMsg::GetMessage { id } => query::get_queued_message(deps.storage, id),
+        QueryMsg::GetDenomRestrictions { denom } => {
+            query::get_denom_restrictions(deps.storage, denom)
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
```

Comment:
why don't we empty the migrate func for simplicity and brevity? it doesn't seem to be applicable for us